File: LintCode/Problems/C++/502. 迷你Cassandra.cpp
/**
 * Definition of Column:
 * class Column {
 * public:
 *     int key;
 *     String value;
 *     Column(int key, string value) {
 *         this->key = key;
 *         this->value = value;
 *    }
 * }
 */


class MiniCassandra {
public:
    MiniCassandra() {
        // do intialization if necessary
    }

    /*
     * @param row_key: a string
     * @param column_key: An integer
     * @param value: a string
     * @return: nothing
     */
    void insert(string &row_key, int column_key, string &value) {
        // write your code here
        this->dic[row_key][column_key] = value;
    }

    /*
     * @param raw_key: a string
     * @param column_start: An integer
     * @param column_end: An integer
     * @return: a list of Columns
     */
    vector<Column> query(string &row_key, int column_start, int column_end) {
        // write your code here
        std::vector<Column> result;
        std::map<int, std::string>& tmp = this->dic[row_key];
        for (auto i = tmp.begin(); i != tmp.end(); ++i) {
            if (i->first >= column_start && i->first <= column_end) {
                result.push_back(Column(i->first, i->second));
            }
        }
        return result;
    }

private:
    std::map<std::string, std::map<int, std::string>> dic; // map is sorted map
};

File: LintCode/Problems/Python3/502. 迷你Cassandra.py
# column_key should using OrderDict
# Your submission beats 29.67% Submissions!
# Total runtime 1573 ms


"""
Definition of Column:
class Column:
    def __init__(self, key, value):
        self.key = key
        self.value = value
"""


class MiniCassandra:

    def __init__(self):
        # do intialization if necessary
        self.dic = {}

    """
    @param: row_key: a string
    @param: column_key: An integer
    @param: value: a string
    @return: nothing
    """
    def insert(self, row_key, column_key, value):
        # write your code here
        if row_key in self.dic:
            for item in self.dic[row_key]:
                if item[0] == column_key:
                    self.dic[row_key].remove(item)
                    break
            self.dic[row_key].append(tuple((column_key, str(value))))
        else:
            self.dic[row_key] = [tuple((column_key, str(value)))]

    """
    @param: row_key: a string
    @param: column_start: An integer
    @param: column_end: An integer
    @return: a list of Columns
    """
    def query(self, row_key, column_start, column_end):
        # write your code here
        if row_key not in self.dic:
            return []
        result = []
        for item in self.dic[row_key]:
            if column_start <= item[0] <= column_end:
                result.append(Column(item[0], item[1]))
        result.sort(key=lambda x: x.key)
        return result
